app/models/agents/website_agent.rb

```diff
@@ -292,11 +304,16 @@ def extract_json(doc)
     end
 
     def extract_text(doc)
-      extract_each(doc) { |extraction_details|
-        regexp = Regexp.new(extraction_details['regexp'])
-        result = []
-        doc.scan(regexp) {
-          result << Regexp.last_match[extraction_details['index']]
+      regexp_cache = {}
+      matches_cache = {}
+      extract_each { |extraction_details|
+        src = extraction_details['regexp']
+        regexp = (regexp_cache[src] ||= Regexp.new(src))
+        matches = (matches_cache[src] ||= [].tap { |ms|
```

Comment:
The part for introducing cache is irrelevant to the main feature of this PR, so I'll leave it for now and just merge the `use_namespaces` option!